### PR TITLE
fix: fix the panic in loading state from storage

### PR DESCRIPTION
**Pull request by utkuozdemir**
When importing clusters from a file, grow the used buffer to the correct size of the next cluster, so that it does not panic when unmarshaling it.

Handle panics on the storage's save & load, so that it will never crash the discovery service when it fails.

Additionally:
- fix the slice growing logic when exporting clusters so that we avoid over-growing the slices for the affiliates and the endpoints.
- modify the storage tests to use the real state instead of a mock, replace the assertions to ignore the order accordingly.

### CODE REVIEW

File: internal/state/snapshot.go [hunk 1]
```diff
@@ -67,7 +67,10 @@ func snapshotCluster(cluster *Cluster, snapshot *storagepb.ClusterSnapshot) {
 	snapshot.Id = cluster.id
 
 	// reuse the same slice, resize it as needed
-	snapshot.Affiliates = slices.Grow(snapshot.Affiliates, len(cluster.affiliates))
+	if len(cluster.affiliates) > cap(snapshot.Affiliates) {
+		snapshot.Affiliates = slices.Grow(snapshot.Affiliates, len(cluster.affiliates)-len(snapshot.Affiliates))
+	}
```
Comment (utkuozdemir) [inline internal/state/snapshot.go:72]: this logic was grow was wrong and causing unnecessary growing.

File: internal/state/snapshot.go [hunk 2]
```diff
@@ -88,7 +91,10 @@ func snapshotCluster(cluster *Cluster, snapshot *storagepb.ClusterSnapshot) {
 		snapshot.Affiliates[i].Data = affiliate.data
 
 		// reuse the same slice, resize it as needed
-		snapshot.Affiliates[i].Endpoints = slices.Grow(snapshot.Affiliates[i].Endpoints, len(affiliate.endpoints))
+		if len(affiliate.endpoints) > cap(snapshot.Affiliates[i].Endpoints) {
+			snapshot.Affiliates[i].Endpoints = slices.Grow(snapshot.Affiliates[i].Endpoints, len(affiliate.endpoints)-len(snapshot.Affiliates[i].Endpoints))
+		}
+
```
Comment (utkuozdemir) [inline internal/state/snapshot.go:97]: same as above

File: internal/state/storage/storage.go [hunk 3]
```diff
@@ -254,7 +268,7 @@ func (storage *Storage) Import(reader io.Reader) (SnapshotStats, error) {
 		}
 
 		if clusterSize > cap(buffer) {
-			buffer = slices.Grow(buffer, clusterSize-cap(buffer))
+			buffer = slices.Grow(buffer, clusterSize)
```
Comment (utkuozdemir) [inline internal/state/storage/storage.go:271]: this is the actual fix for the panic on import

**Review by smira — APPROVED** [review]

**Pull request merged**